ext/ddtrace_profiling_native_extension/collectors_thread_context.c

```diff
@@ -1122,17 +1123,30 @@ void thread_context_collector_sample_allocation(VALUE self_instance, unsigned in
       type == RUBY_T_SYMBOL   ||
       type == RUBY_T_FIXNUM
     ) {
-      const char *name = rb_obj_classname(new_object);
-      size_t name_length = name != NULL ? strlen(name) : 0;
-
-      if (name_length > 0) {
-        class_name = (ddog_CharSlice) {.ptr = name, .len = name_length};
+      VALUE klass = rb_class_of(new_object);
+
+      // Ruby sometimes plays a bit fast and loose with some of its internal objects, e.g.
+      // `rb_str_tmp_frozen_acquire` allocates a string with no class (klass=0).
+      // Thus, we need to make sure there's actually a class before getting its name.
+
+      if (klass != 0) {
+        const char *name = rb_obj_classname(new_object);
+        size_t name_length = name != NULL ? strlen(name) : 0;
+
+        if (name_length > 0) {
+          class_name = (ddog_CharSlice) {.ptr = name, .len = name_length};
+        } else {
+          // @ivoanjo: I'm not sure this can ever happen, but just-in-case
```

Comment:
Is it possible for this to happen with anonymous classes?

Reply:
Good question! It shouldn't happen -- we get the Ruby autogenerated name for them:

```ruby
[1] pry(main)> Class.new.to_s
=> "#<Class:0x00005b721ef2a558>"
```

Here's the output of an actual profile with such a class:

```
allocation class:[#<Class:0x00005dd2c7da98a8>] ruby vm type:[T_OBJECT] thread id:[123612321531712 (2560)] thread name:[main]
```